userchroot.c

```diff
@@ -566,11 +566,28 @@ int main(int argc, char* argv[], char* envp[]) {
       argv[2] != NULL &&
       argv[2][0] == '-') {
 
-    // this mode can only be run by the owner of the chroot image.
-    if (target_user != statbase_path.st_uid) {
-      fprintf(stderr,"install or uninstall devices can only be called by the owner of the chroot. Aborting.\n");
+    int dev_path_size = strlen(final_path) + strlen("/dev") + 1;
+    char* dev_path = malloc(dev_path_size);
+    if (dev_path == NULL) {
+      fprintf(stderr,"Failed to allocate memory. Aborting.\n");
       exit(ERR_EXIT_CODE);
     }
+    rc = snprintf(dev_path, dev_path_size, "%s/dev", final_path);
+    if (rc <= 0) {
+      fprintf(stderr,"Failed to assemble /dev path. Aborting.\n");
+      exit(ERR_EXIT_CODE);
+    }
+
+    // this mode can only be run by users with write access to the
+    // /dev directory in the chroot image.
+    if (access(dev_path, R_OK | W_OK | X_OK) != 0) {
```

Comment:
In addition to doing this, does it make sense to always allow the owner of the chroot to call install/uninstall devices? I can't quite imagine a case in practice where the owner of the chroot wouldn't have write permissions to `/dev`, but if such a case came up that's the behavior I'd expect.

Reply:
If this was a new tool, I would probably not add such an exception unless there was a clear use case. However, to reduce the risk of breaking scenarios for current users of `userchroot` it sounds sensible.

If we make sure that `/dev` is not a symlink, always allowing the chroot owner to install/uninstall devices should be ok. If `/dev` was a symlink, the chroot owner could get `userchroot` to manipulate the wrong directory.

Side-note: All these path checks before the real operations are theoretically susceptible to small TOCTOU race conditions. However, that's already the case in master, so those should be fixed in a separate PR if considered significant. Directory fds and *at() syscalls should help if they are supported by all supported platforms.

Reply:
I've updated the branch with the `/dev` symlink check and allowing chroot owners to install devices even if they don't have write access to `/dev`.